internals/api/auth.go

```diff
@@ -0,0 +1,249 @@
+package api
+
+import (
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/secrethub/secrethub-go/internals/api/uuid"
+)
+
+// AuthMethod options
+const (
+	AuthMethodAWSSTS = "aws-sts"
+)
+
+// SessionType options
+const (
+	SessionTypeHMAC SessionType = "hmac"
+)
+
+// Errors
+var (
+	ErrInvalidSessionType  = errAPI.Code("invalid_session_type").StatusError("invalid session type provided for authentication request", http.StatusBadRequest)
+	ErrInvalidPayload      = errAPI.Code("invalid_payload").StatusError("invalid payload provided for authentication request", http.StatusBadRequest)
+	ErrInvalidAuthMethod   = errAPI.Code("invalid_auth_method").StatusError("invalid auth method", http.StatusBadRequest)
+	ErrMissingField        = errAPI.Code("missing_field").StatusErrorPref("request is missing field %s", http.StatusBadRequest)
+	ErrSessionNotFound     = errAPI.Code("session_not_found").StatusError("session could not be found, it might have expired", http.StatusForbidden)
+	ErrSessionExpired      = errAPI.Code("session_expired").StatusError("session has expired", http.StatusForbidden)
+	ErrAuthFailed          = errAPI.Code("auth_failed").StatusError("authentication failed", http.StatusForbidden)
+	ErrCouldNotGetEndpoint = errAPI.Code("wrong_endpoint").StatusError("could not find the AWS endpoint", http.StatusBadRequest)
+	ErrAWSException        = errAPI.Code("aws_exception").StatusError("AWS returned an error", http.StatusFailedDependency)
+	ErrNoServiceWithRole   = errAPI.Code("no_service_with_role").StatusErrorPref("there exists no service account tied to the AWS IAM role '%s'", http.StatusNotFound)
+)
+
+// SessionType defines how a session can be used.
+type SessionType string
+
+// AuthRequest is a request to authenticate and request a session.
+type AuthRequest struct {
+	Method      *string      `json:"method"`
+	SessionType *SessionType `json:"session_type"`
```

Comment:
For unmarshaling it makes sense. This way you can detect if a field is not set instead of empty. `omitempty` has no effect on this. So if someone uses our API without using our client, they get more usable error messages.